File: python/363_矩形区域不超过K的最大数值和_hard.py
from sortedcontainers import SortedList
class Solution:
    def maxSumSubmatrix(self, matrix: List[List[int]], k: int) -> int:
        ans = float("-inf")
        m, n = len(matrix), len(matrix[0])
        for i in range(0, m):
            d = [0]*n
            for j in range(i, m):
                for c in range(0, n):
                    d[c] += matrix[j][c]
                sumSet = SortedList([0])
                s = 0
                for v in d:
                    s += v
                    lb = sumSet.bisect_left(s - k)
                    if lb != len(sumSet):
                        ans = max(ans, s - sumSet[lb])
                    sumSet.add(s)
        return ans

File: cpp/363_矩形区域不超过K的最大数值和_hard.cpp
class Solution {
public:
    int maxSumSubmatrix(vector<vector<int>>& matrix, int k) {
        int m = matrix.size(), n = matrix[0].size();
        int ans = INT_MIN;
        for (int i = 0; i < m; i++) {
            vector<int> d(n);
            for (int j = i; j < m; j++) {
                for (int c = 0; c < n; c++) {
                    d[c] += matrix[j][c];
                }
                set<int> sumSet{0};
                int s = 0;
                for (auto v : d) {
                    s += v;
                    auto lb = sumSet.lower_bound(s-k);
                    if (lb != sumSet.end())
                        ans = max(ans, s - *lb);
                    sumSet.insert(s);
                }
            }
        }
        return ans; 
    }
};